scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,199 @@
+Doing this solves both the remaining aspects of the re-indexing problem and the database ever-growing problem,
+as unneeded Index Reports are permanently deleted and needed reports may just be regenerated.
+
+**Note**: this approach also solves the "versioned scanner" update scenario; however, we opt to implement the [`checkManifest`](#reuse-claircores-checkmanifest-step)
+step for faster turnaround upon "versioned scanner" updates.
+
+Of course, determining a timeframe for deletion is non-trivial. Deleting too quickly means Index Reports may be corrected
+faster, but too many (probably) unnecessary requests to the image registry and (probably) unnecessary CPU/memory/disk usage.
+Deleting too slowly means resource usage will be much less, but it may take (too) long to correct the Index Report.
+Similarly, we would need to ensure not all Index Reports are all re-indexed at the same time (hence random deletions)
+to ensure spikes are minimized.
+
+The deletions will be chosen from a range between **one week** and **one month**. We believe this is a sufficient timeframe
+which balances speed and resource consumption.
+
+This will all be tracked in a new table (see [Manifest Metadata Table](#manifest-metadata-table) below).
+
+The table will be queried periodically for passed timestamps. These rows will be deleted and the related Index Reports will be deleted from ClairCore's tables.
+
+#### Alternative - Client Delete Requests
+
+Scanner V4 clients (Central and Sensor) may inform the Indexer of an image deletion and request the Index Report deleted
+
+However:
+
+* It is non-trivial to pinpoint all places where images are deleted in Central and Sensor
+* The proposed method is simpler, as it is all done from within the Indexer
+
+### Manifest Metadata Table
+
+The new Manifest Metadata table will look like the following:
+
+| manifest_id   | delete_after                  |
+|---------------|-------------------------------|
+| sha512:abc... | 2024-09-26 23:52:39.190285-07 |
+| sha512:def... | 2024-09-12 05:48:46.361476-07 |
+
+The `manifest_id` will be all that is needed to identify and delete an Index Report. Note this ID is the same as the
+Index Report's ID which is the same as the related `*claircore.Manifest.Hash` (hence `manifest_id`).
+
+### Handling preexisting Index Reports
+
+This requires a PostgreSQL migration. This migration will act as if the Index Reports were generated upon startup, as follows:
+
+```sql
+--- a table to hold deletion timestamps for manifests.
+CREATE TABLE IF NOT EXISTS manifest_metadata (
+    manifest_id  TEXT PRIMARY KEY,
+    delete_after TIMESTAMP
+);
+
+--- add all manifests from the manifest table to the manifest_metadata table
+--- if they do not already exist in the manifest_metadata table.
+---
+--- set the delete_after column to some random time between seven and thirty days from now.
+INSERT INTO manifest_metadata (manifest_id, delete_after)
```

Comment:
Edit: I will be removing this migration in favor of doing this in Go, as the migration would be more complex than originally expected. We'd need to ensure the `manifest` table remains as-is for our migration to work properly. The only way to guarantee it is to inject this migration into ClairCore's Indexer migration in the current place. This is possible to do correctly (I believe), but not ideal. We should not really mess with ClairCore's internal migrations. Instead, it may be better/cleaner to do this post-migrations as one of the first things done upon startup in Go.

We already have precedence of running a single GC cycle at a time, so this would be similar.